### PR TITLE
Take a pass at fixing metrics API javadoc issues/missing information.

### PR DESCRIPTION
**Pull request by jsuereth**
- Add "how to select instrument" guidance to `Meter`.
- Add information about `bound` and its limitations
- Add expectations of callbacks.
- Improve documentation on `UpDownCounter` vs. `Counter`.

### CODE REVIEW

**Review by bogdandrutu — APPROVED** [review]

File: api/metrics/src/main/java/io/opentelemetry/api/metrics/package-info.java [hunk 1]
```diff
@@ -3,7 +3,17 @@
...
+/**
+ * This package describes the Metrics API that can be used to record application Metrics.
+ *
+ * <p>The primary entry point to Metrics is the {@link io.opentelemetry.api.metrics.MeterProvider},
+ * which allows the construction of a {@link io.opentelemetry.api.metrics.Meter}. Instrumentated
+ * libraries should construct a single {@link io.opentelemetry.api.metrics.Meter} and register
+ * `instruments` via the builders available on {@link io.opentelemetry.api.metrics.Meter}.
+ *
+ * <p>There is a global instance of {@link io.opentelemetry.api.metrics.MeterProvider} available for
+ * scenarios where instrumentation authors are unable to obtain one by other means.
```
Comment (jkwatson) [inline api/metrics/src/main/java/io/opentelemetry/api/metrics/package-info.java:15]: I wonder if it's worth documenting this until we have put metrics into the main OpenTelemetry instance, and GlobalOpenTelemetry.

File: api/metrics/src/main/java/io/opentelemetry/api/metrics/LongCounterBuilder.java [hunk 2]
```diff
@@ -41,6 +41,14 @@
    *
    * <p>The callback will only be called when the {@link Meter} is being observed.
    *
+   * <p>Callbacks are expected to abide by the following restrictions:
+   *
+   * <ul>
+   *   <li>Run in a finite amount of time.
+   *   <li>Safe to call repeatedly, across multiple threads.
+   *   <li>Return positive, monotonically increasing values.
```
Comment (jack-berg) [inline api/metrics/src/main/java/io/opentelemetry/api/metrics/LongCounterBuilder.java:49]: Not related to the documentation, but should the SDK be defensive against the callbacks throwing, and wrap them in a try / catch?

**Review by jack-berg — APPROVED** [review]

File: api/metrics/src/main/java/io/opentelemetry/api/metrics/LongCounterBuilder.java (hunk 2 above)
Comment (jsuereth) [reply api/metrics/src/main/java/io/opentelemetry/api/metrics/LongCounterBuilder.java:49]: Yes, can you open a bug for that?

Comment (jack-berg) [reply api/metrics/src/main/java/io/opentelemetry/api/metrics/LongCounterBuilder.java:49]: Issue #3891.

File: api/metrics/src/main/java/io/opentelemetry/api/metrics/package-info.java (hunk 1 above)
Comment (jkwatson) [reply api/metrics/src/main/java/io/opentelemetry/api/metrics/package-info.java:15]: I'm ok either way, as long as we don't forget. ;)

**Review by jkwatson — APPROVED** [review]

**Pull request merged**